src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+                        if (symbolValueUsageStateMap.TryRemove(member, out var valueUsageInfo) &&
+                            (valueUsageInfo & (ValueUsageInfo.Read | ValueUsageInfo.ReadableRef)) == 0)
+                        {
+                            Debug.Assert(IsCandidateSymbol(member));
+
+                            var option = GetCodeStyleOption(member, symbolEndContext);
+                            if (option != null && option.Value)
+                            {
+                                // Give appropriate message based on usage:
+                                //  1. No read or writes:   Type '{0}' has an unused private member '{1}' which can be removed.
+                                //  2. Only writes:         Type '{0}' has a private member '{1}' which can be removed as the value assigned to it is never used.
+
+                                var rule = (valueUsageInfo & (ValueUsageInfo.Write | ValueUsageInfo.WritableRef)) != 0 ?
+                                    _noReadsUnnecessaryWithSuggestionDescriptor :
+                                    UnnecessaryWithSuggestionDescriptor;
+
+                                var additionalLocations = member.Locations.Skip(1).ToImmutableArrayOrEmpty();
+                                var diagnostic = DiagnosticHelper.Create(
+                                    rule,
+                                    member.Locations[0],
+                                    option.Notification.Severity,
+                                    additionalLocations,
+                                    properties: ImmutableDictionary.CreateRange(new[] {
+                                        new KeyValuePair<string, string>(UnunsedMemberNameProperty, member.Name),
+                                        new KeyValuePair<string, string>(UnunsedMemberKindProperty, member.Kind.ToString()) }),
+                                    member.ContainingSymbol.Name,
+                                    member.Name);
+                                symbolEndContext.ReportDiagnostic(diagnostic);
+                            }
+                        }
+                    }
+                }
+
+                bool IsCandidateSymbol(ISymbol memberSymbol)
+                {
+                    if (memberSymbol.DeclaredAccessibility == Accessibility.Private &&
+                        !memberSymbol.IsImplicitlyDeclared)
+                    {
+                        // Do not track accessors, as we will track the associated symbol.
+                        switch (memberSymbol.Kind)
+                        {
+                            case SymbolKind.Method:
+                                return ((IMethodSymbol)memberSymbol).AssociatedSymbol == null &&
+                                    memberSymbol != entryPoint;
+
+                            case SymbolKind.Field:
+                                return ((IFieldSymbol)memberSymbol).AssociatedSymbol == null;
```

Comment:
comments would be good.

Reply:
I have a comment above that we don't intend to track symbols if they have an associated symbol.

---
In reply to: [212734153](https://github.com/dotnet/roslyn/pull/29511#discussion_r212734153) [](ancestors = 212734153)